### PR TITLE
Inform customer about quantity in cart

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Displays how many pieces of product the customer has in his cart on FO product page. Both single and included in packs.
| Type?             | new feature
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | 
| UI Tests          | https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/6797167149
| Fixed issue or discussion?     | Fix https://github.com/PrestaShop/PrestaShop/discussions/32439
| Related PRs       | 
| Sponsor company   | 

### Screenshots
Snímek obrazovky 2023-09-13 174214
Snímek obrazovky 2023-09-13 145925

### CODE REVIEW

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -576,6 +580,56 @@ protected function getMinimalProductOrDeclinationQuantity($product)
...
+        if ($cartProductQuantity['quantity'] != $cartProductQuantity['deep_quantity']) {
+            $quantityInPacks = (int) $cartProductQuantity['deep_quantity'] - (int) $cartProductQuantity['quantity'];
+        }
+
+        // Render nice notifications so the user knows what is happening
+        if ($standaloneQuantity > 0 && $quantityInPacks > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product in cart and another %2s piece(s) included in packs.',
+                [$standaloneQuantity, $quantityInPacks],
+                'Shop.Notifications.Error'
```
Comment (jolelievre) [inline controllers/front/ProductController.php]:
```suggestion
                'Shop.Theme.Catalog'
```

Comment (jolelievre) [reply controllers/front/ProductController.php]: or
```suggestion
                'Shop.Notifications.Success'
```

Comment (Hlavtox) [reply controllers/front/ProductController.php]: Done

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -576,6 +580,56 @@ protected function getMinimalProductOrDeclinationQuantity($product)
...
+        // Render nice notifications so the user knows what is happening
+        if ($standaloneQuantity > 0 && $quantityInPacks > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product in cart and another %2s piece(s) included in packs.',
+                [$standaloneQuantity, $quantityInPacks],
+                'Shop.Theme.Catalog'
+            );
+        } elseif ($standaloneQuantity > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product in cart.',
```
Comment (MatShir) [inline controllers/front/ProductController.php]: Piece feels weird ! Maybe item, quantity (hard to make it plural as (s)) or unit ?

Comment (Hlavtox) [reply controllers/front/ProductController.php]: It's doable for v9 if you push this topic - https://github.com/PrestaShop/PrestaShop/pull/32940/files (https://github.com/PrestaShop/PrestaShop/issues/15870)

**Comment by l-delin** [comment]
Hello @Hlavtox ! 👋
I fail to understand the purpose of this message. Does it result from any need expressed by customers?
I mean, if the customer wants to check how much of a product they've got, they can simply check their cart. Seems to add an unnecessary feature with yet more information in the FO. Maybe I'm missing something. Let me know 🙂

**Comment by Hlavtox** [comment]
@l-delin It's validated by product council. It's mitigating the confusion in front office when customer sees that the product is not in stock, because he has it in cart. And he has no way to know unless he acually goes to the cart page.

This way he will see that he has this product in his cart.

> Seems to add an unnecessary feature with yet more information in the FO.

There is never ever ever ever too much information in FO.

**Comment by MatShir** [comment]
I will add that the feature can be disabled. The theme developer can personalize the experience as they wish also. We need to be able to provide all the resource for the dev & merchants.

**Comment by MatShir** [comment]
Here the GH [discussions](https://github.com/PrestaShop/PrestaShop/discussions/32439) and [slack convo](https://prestashop.slack.com/archives/C03R7N8G28H/p1694606995359959), it is a shared issue by different contributors.

**Comment by MatShir** [comment]
But I don't see the switch that might explain the confusion @Hlavtox 🤔

**Comment by Hlavtox** [comment]
@MatShir WIP

**Comment by Hlavtox** [comment]
@jolelievre @MatShir Made it to work properly with products with combinations and customizations. Switch added. :-)

File: controllers/front/ProductController.php [hunk 3]
```diff
@@ -576,6 +580,60 @@ protected function getMinimalProductOrDeclinationQuantity($product)
...
+
+        // If we have some products in pack only, we will subtract it to
+        if ($cartProductQuantity['quantity'] != $cartProductQuantity['deep_quantity']) {
+            $quantityInPacks = (int) $cartProductQuantity['deep_quantity'] - (int) $cartProductQuantity['quantity'];
+        }
+
+        // Render nice notifications so the user knows what is happening
+        if ($standaloneQuantity > 0 && $quantityInPacks > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product in cart and another %2s piece(s) included in packs.',
```
Comment (l-delin) [inline controllers/front/ProductController.php]:
```suggestion
                'Your cart contains %1s (product name), and %2s (product name) are included in packs.',
```

Comment (l-delin) [reply controllers/front/ProductController.php]: Is it possible to add something like but replace the (product name) comment with a variable that fetches the name of the current product displayed on page?

File: controllers/front/ProductController.php [hunk 4]
```diff
@@ -576,6 +580,60 @@ protected function getMinimalProductOrDeclinationQuantity($product)
...
+                [$standaloneQuantity, $quantityInPacks],
+                'Shop.Theme.Catalog'
+            );
+        } elseif ($standaloneQuantity > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product in cart.',
```
Comment (l-delin) [inline controllers/front/ProductController.php:624]:
```suggestion
                'Your cart contains %1s (product name).',
```

File: controllers/front/ProductController.php [hunk 5]
```diff
@@ -576,6 +580,60 @@ protected function getMinimalProductOrDeclinationQuantity($product)
...
+                [$standaloneQuantity],
+                'Shop.Theme.Catalog'
+            );
+        } elseif ($quantityInPacks > 0) {
+            $this->info[] = $this->trans(
+                'You have %1s piece(s) of this product included in packs in your cart.',
```
Comment (l-delin) [inline controllers/front/ProductController.php:630]:
```suggestion
                'Your cart contains %1s (product name) included in packs.',
```

Comment (l-delin) [inline controllers/front/ProductController.php]: same

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 6]
```diff
@@ -97,6 +97,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
                 ),
                 'required' => false,
+            ])
+            ->add('display_amount_in_cart', SwitchType::class, [
+                'label' => $this->trans(
+                    'Display notifications if the product is already in the cart',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'If the customer already has this product in his cart, a notification will be displayed on top of the product page. The customer will also be notified if the product he is browsing is part of a pack he has in his cart.',
```
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]:
```suggestion
                    'If your customer already has this product in their cart, a notification will be displayed on top of the product page. The customer will also be notified if the product they are browsing is part of a pack that's in their cart.',
```

Comment (l-delin) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: Some edits for inclusivity :-)

**Review by l-delin — COMMENTED** [review]
Some changes need to be made. Let me know regarding the product name variables 🙂

File: controllers/front/ProductController.php (hunk 3 above)
Comment (Hlavtox) [reply controllers/front/ProductController.php]: I don't think it's required to show the product name if the message is related to the product the customer is viewing right now. 👍Plus your wording is weird imho.

Comment (MatShir) [reply controllers/front/ProductController.php]: Product name can be very long. I think also it adds some complexity in the warning

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 6 above)
Comment (Hlavtox) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: @l-delin Applied ;-)

File: controllers/front/ProductController.php (hunk 3 above)
Comment (Hlavtox) [reply controllers/front/ProductController.php]: Another problem is other languages. We have different word endings for each grammatical case + plurals.

English: Your cart contains 5 Hummingbird printed sweater, and 1 Hummingbird printed sweater are included in packs.
Czech: Váš košík obsahuje 5 Hummingbird printed sweaterů a 1 Hummingbird printed sweater máte v balíčcích.

It looks bad even in English when you don't have the plural.

Comment (axelvrsc) [reply controllers/front/ProductController.php]: Hi, 
Thanks @l-delin @Hlavtox for your propositions. 
As we're trying to be specific in a context where we just add complementary informations to avoid friction for users and avoid them to check their cart directly it's only fair that we're also being specific about what we talk about.

If we announce a quantity and we know a quantity of what, then there's a sense in adding it in the formulation rather than being vague with "product", even though we could definitely argue that users know what product we're talking about.

As content relies on grammar, syntax, semantics and context, let's try not to evaluate content based on feeling, such as it being weird or not, but rather evaluate its efficiency regarding a specific goal. 

On a language basis, "a piece of product" doesn't necessarily design a unit but rather a part of a product (as if you'd separate it).
 So we'll go with something that is at least correct in english. 
 
 Thanks !

Comment (MatShir) [reply controllers/front/ProductController.php]: We all agree that the "piece" is not ideal wording, but I don't think using the product name will be suitable in this context.

Comment (Hlavtox) [reply controllers/front/ProductController.php]: @axelvrsc 

> If we announce a quantity and we know a quantity of what, then there's a sense in adding it in the formulation rather than being vague with "product", even though we could definitely argue that users know what product we're talking about.

If the user is viewing the product page and the message says **YOUR CART CONTAINS 2 PIECES OF THIS PRODUCT**, then the user must be an complete idiot if he doesn't know what we are talking about. 👍

> As content relies on grammar, syntax, semantics and context, let's try not to evaluate content based on feeling, such as it being weird or not, but rather evaluate its efficiency regarding a specific goal.

My "feeling" is based on working in ecommerce for every day for 15 years now + studying translation. :-)

> So we'll go with something that is at least correct in english.

We must go with something that is correct in all languages, if we don't have a proper way to ICU translate strings.

Comment (MatShir) [reply controllers/front/ProductController.php]: One example of product name: "PC Portable Lenovo IdeaPad 3 15ALC6 15,6" AMD Ryzen 5 8 Go RAM 256 Go SSD Gris arctique"

Comment (Hlavtox) [reply controllers/front/ProductController.php]: piece noun
 
- a single object of a particular type:
- a piece of furniture/clothing/equipment
- a piece of paper (= a whole sheet)
- a piece of china (= an object made of china)
- a piece of information/advice

Comment (MatShir) [reply controllers/front/ProductController.php]: An update:
I will push the review anyway to have the feature, hopefully for 8.1.2. In parallel, I'm planning a quick discussion with @axelvrsc so we can find a suitable candidate. If in the meantime, the feature is merged. We can still improve it later.

Comment (Hlavtox) [reply controllers/front/ProductController.php]: pay/buy/sell by the piece - COMMERCE to pay for, buy, or sell something in single units rather than in groups

Comment (MatShir) [reply controllers/front/ProductController.php]: > If the users is viewing the product page and the message says YOUR CART CONTAINS 2 PIECES OF THIS PRODUCT, the user must be an complete idiot if he doesn't know what we are talking about. 👍

"Piece" can be used on too many different contexts, even translator could get it wrong.

Comment (axelvrsc) [reply controllers/front/ProductController.php]: Thanks @MatShir for coming with a proper argument, that's a good reason not to use the product name indeed. 

@Hlavtox Intuition from experience is always good, I can't deny that, but that cannot be a source of truth. We're trying to convey an information for a specific context as you clearly identified and thanks again for that! That's exactly the reason why we need to find the best way to indicate it.

As I said, we can't talk about a piece of an article as the article is not a sum of something that we're dividing. We're trying to find a way to talk about the numbers of unit of the selected product. It's also very important for translating in other languages as you just said because quantities and qualities are very specific according to certain languages. 

We're all content specialists here, you're not the only one with a background in translation and experience, there's no need to get against each other as we all share the same goal of improving the Open Source project and providing users with the best experience :)

Comment (MatShir) [reply controllers/front/ProductController.php]: @Hlavtox, FYI: @axelvrsc didn't introduce himself yet, but he will take over @l-delin works.

Comment (Hlavtox) [reply controllers/front/ProductController.php]: @MatShir Oh OK, I wrote it thinking somebody is putting his nose into things! :D

**To summarize:**
- Product name would be nice sometimes, but we are unable to do it properly, even in English.  It can be long and ugly. It would require joining some more tables.
- We have a czech word "kus"/"kusy" which means how many articles are there to purchase in this context. The direct translation is "pieces" and I didn't find any other english word that would mean the same thing. I posted some things above. :-)

Comment (Hlavtox) [reply controllers/front/ProductController.php]: My original wording is the best I think 🤷

Snímek obrazovky 2023-09-14 151811

Comment (MatShir) [reply controllers/front/ProductController.php]: Option 3 appears more aesthetically pleasing. It addresses the concern regarding the "piece" and is in proximity to the success message, aligning with the overall context.

Comment (l-delin) [reply controllers/front/ProductController.php]: Hello, yes. I think ultimately we can use the following as a success message on the FO product page:

"You've successfully added %1s products to your cart."
"You've successfully added %1s products included in packs in your cart."
"You've successfully added %1s products and %2s products included in packs in your cart."

Any improvement can be made later. Hopefully your ICU can fix the product/products issue so we don't have to worry about it anymore @Hlavtox

Comment (axelvrsc) [reply controllers/front/ProductController.php]: You're right @MatShir sorry @Hlavtox I forgot to introduce myself, my mistake! 
I'm Axel, Lea and Julie's manager and will be working with Julie on the wording contributions now that Lea will not be working with us from tomorrow. 

Thanks for your exchanges and the details you put. 
I'm glad we found a first wording to start with. As @l-delin said we'll see later if improvement is needed.

Thanks again @Hlavtox for your contribution, looking forward to work on other wordings with you!

Comment (Hlavtox) [reply controllers/front/ProductController.php]: @l-delin It's not a success message, it's displayed after you revisit the product page. `You've successfully added bla bla` doesn't really fit.

This is how it works:
https://github.com/PrestaShop/PrestaShop/assets/6097524/f197cf6c-7d91-4db8-b5dd-5dac3bd36b79

Comment (Hlavtox) [reply controllers/front/ProductController.php]: If you want to avoid the pieces, I have thought of this now:
- Your cart contains 1 of these products.
- Your cart contains 5 of these products.
- Your cart contains 5 of these products and another 5 are included in packs in your cart.
- 5 of these products are included in packs in your cart.

WDYT???

Comment (l-delin) [reply controllers/front/ProductController.php]: Hey @Hlavtox I like the last proposition. We can go with this.

**Comment by l-delin** [comment]
Wording ✅ Thanks

**Review by matks — DISMISSED** [review]
Woah guys this PR is meddling with the cart quantity calculations and this part has no unit test.

No way I approve this PR to go inside a patch release this stuff is way too fragile and prone to errors. Patch releases aim to strengthen the software and fix bugs, risky stuff like this is for minor/major versions that benefit from a feature freeze and beta period to catch and stabilize such sensible topics.

**Comment by matks** [comment]
(sorry to be the party pooper) 😅 I promise I do this in the best interest of the project

**Comment by kpodemski** [comment]
This time I agree with @matks 

@Hlavtox 

We need to be careful, it's the same as with images. You want to fix something? OK, let's fix and leave improvement for another PR. If it improves something but it's a bit more complicated, let's try to not rush things a few days before release.

This doesn't fix anything as far as I understand the issue, it changes confusing behavior - that's something we can live with if we lived for a few years. I fully support the intend and improvement tho!

**Review by jolelievre — DISMISSED** [review]
I agree with @matks you went way too far compared to the initial PR and the purpose that was mearly to add a label to clarify the content of the cart

Not only did it become a feature PR but it also modified some functions that are very unstable and yet critical

**Comment by Hlavtox** [comment]
@matks @kpodemski @jolelievre  Guys, I have removed the modifications from other method. Now it doesn't change anything and is completely safe.

It's working perfectly for products with combinations and customizations.

File: controllers/front/ProductController.php [hunk 7]
```diff
@@ -576,6 +580,60 @@ protected function getMinimalProductOrDeclinationQuantity($product)
         return $minimalProductQuantity;
     }
 
+    /**
+     * Displays information, if the customer has this product in cart already.
+     */
+    protected function addCartQuantityNotification()
+    {
+        if ((bool) Configuration::get('PS_DISPLAY_AMOUNT_IN_CART') !== true) {
```
Comment (jolelievre) [inline controllers/front/ProductController.php:572]: Is it really necessary to configure this? 

Why not simply show this all the time and remove this new configuration?

If it were a more critical change I'd advise adding a configuration for backward compatibility and all, but in this case it seems overkill

Comment (Hlavtox) [reply controllers/front/ProductController.php:572]: Because not all people may want to enable this behavior, and given it's in the core, it's not something where we can say “modify your theme”. For example, some people may not like that it doesn't refresh automatically. 

The change is safe, even if there is no config in database, it will work properly. 👍

Comment (jolelievre) [reply controllers/front/ProductController.php:572]: Ok but then I think more and more that this PR is a new feature and it probably should target develop branch then, sorry 😅

Comment (Hlavtox) [reply controllers/front/ProductController.php:572]: I think we said we can add new features into patches if its safe no? 👀

Comment (kpodemski) [reply controllers/front/ProductController.php:572]: and absolutely necessary :)

**Review by kpodemski — DISMISSED** [review]
Let's move it to v9.

**Review by matthieu-rolland — COMMENTED** [review]
Let's change the base to develop @Hlavtox 👍

**Comment by Murat-06** [comment]
> @l-delinÜrün konseyi tarafından onaylanmıştır. Müşterinin sepetinde ürün olduğu için stokta olmadığını görmesi ön bürodaki karışıklığı azaltıyor. Ve gerçekten sepet sayfasına gitmediği sürece bunu bilmesinin hiçbir yolu yok.
> 
> Böylece sepetinde bu ürünün bulunduğunu görecektir.
> 
> > FO'ya daha fazla bilgi içeren gereksiz bir özellik eklenmiş gibi görünüyor.
> 
> FO'da asla çok fazla bilgi yoktur.

**Comment by Murat-06** [comment]
When a product with 1 item left in stock is added to the cart, it immediately appears that the product is out of stock. This is a very bad situation. When the customer comes later, he leaves the site saying it is out of stock, unaware that the product is in his cart.

**Comment by Hlavtox** [comment]
@kpodemski @matthieu-rolland @jolelievre Rebased onto develop 👍

**Review by kpodemski — DISMISSED** [review]

**Comment by M0rgan01** [comment]
@Hlavtox I took the liberty of rerunning the UI tests on develop

**Review by M0rgan01 — DISMISSED** [review]

**Review by nesrineabdmouleh — DISMISSED** [review]

**Comment by AureRita** [comment]
Hi @Hlavtox 

Thank you for your PR, I tested it and : 

This PR doesn't count the number of product in a pack as you can see : 

[Untitled_ Oct 17, 2023 3_57 PM.webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/baa193f9-ff15-4882-a700-fb3d73830ac5)

Thank you

**Comment by MatShir** [comment]
@AureRita on your video, it is hard to follow what is failing or not. Could share one detail the failing use case, please ? So @Hlavtox can reproduce it on his end 🙏

**Review by M0rgan01 — APPROVED** [review]

**Comment by Hlavtox** [comment]
@AureRita You can retest it :-)

**Review by AureRita — APPROVED** [review]
Hi @Hlavtox 

Thank you for your PR, I tested it and it seems to works as you can see : 

[Untitled_ Nov 9, 2023 4_48 PM.webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/96941bc7-cc8c-4498-baac-ee3de544cca1)

Link to the auto test : https://github.com/AureRita/testing_pr/actions/runs/6812892397

Because the auto test is :green_circle: and the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you
